### ISSUE TITLE
Would Onyx be suited for polyglot programming?

### ISSUE THREAD

**Issue by joelkuiper** [onyx-platform/onyx#193]
I'm sorry if this has been answered before, but I was wondering if Onyx would be useful for distributed computation in other languages as well?

Occasionally I need to interface with existing Python, Javascript and R libraries. Currently I'm using my own makeshift RPC framework for this, but I'd love to move to something more supported&robust. 

Since Onyx functions are just regular Clojure functions, would it make sense to do some child process work or shell calls in them? 

Thanks!

**Comment by MichaelDrogalis** [comment]
Hi @joelkuiper, great question.

I'll answer your last question first - you can absolutely do shell calls and spawn processes in your Onyx tasks. There wouldn't be any extra features that need to be added to Onyx to let you do that.

To answer a little more broadly though, it would be *nice* if you didn't need to resort to shell calls yourself. We're trying to wrap up the `0.6.0` release in the next few weeks, and proper multi-language support is one of the next major things that we're after. You can imagine a catalog with a task that looks something like this:

```clojure
{:onyx/name :inc
 :onyx/lang :python
 :onyx/fn :my.fully.qualified.python.path/function-name
 :onyx/type :function
 :onyx/batch-size batch-size}
```

If we can achieve that, we'd have the ability to mix languages *in the same job*. That'd be truly first class support for situations like you're in. cc @lbradstreet 

In short, yes you can do it, and we're trying to make it even more pleasant. :)

**Comment by joelkuiper** [comment]
Hey @MichaelDrogalis,

Thanks for the quick reply, and this is great news!  Looking forward to the 0.6.0 release. 

Some of the analyses I plan to run might take a while (in the order of minutes). They can't really be broken up either (they're Markov chain Monte Carlo (MCMC) simulations). Would this require any special attention, or will this work "out of the box"? Thanks so much for this wonderful package!

**Comment by MichaelDrogalis** [comment]
Onyx acts as both a batch and a stream processor, and is design for jobs that last on the order of minutes/hours/days for batch loads, and jobs that never come down for streaming loads. You can use the built-in job schedulers to do this. Specifically, you'd use the Greedy job scheduler to ensure that all nodes in your cluster work a single job until it's complete.

If you'd like more parallelism, you can use `:onyx/max-peers` on your catalog entries and bound the number of peers that works on each job. That way you could guarantee each job has a certain amount of peers, but the job isn't "broken up", so to speak. :)

**Comment by joelkuiper** [comment]
Thank you for all the replies! 
Looking forward to integrating this software in our toolsuite! 

Thanks again :smile:

**Comment by jmorris0x0** [comment]
Hi, it's been a while since the original question. What is the state of multi-language support? In `onyx/src/onyx/spec.cljc` I see this:

`(s/def :onyx/language #{:clojure :java})`

...but nothing except Clojure and Java. If not already implemented, what would be required to make this work?

**Comment by MichaelDrogalis** [comment]
Hi @jmorris0x0! Feels like stepping into a time machine looking this thread.

Ultimately we decided that multi-language support wasn't high enough of a priority given everything else we tackled over the years. Clojure and Java remain the only supported runtime environments.

The critical thing to making multi-language support viable in a performant manner is setting up a local channel of communication between a JVM Onyx peer and a process in the target language serves requests. This approach avoids reimplementing the vast majority of the runtime. The trick here is keeping things fast - going back and forth between languages constantly has high serialization overhead.

If you want to jump in Clojurians Slack, we can chat more about the work slated for this ticket.
